### PR TITLE
O3-602: Adding multiple identifiers to a patient

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

This work allows a patient to have multiple identifiers, which can be added, updated and deleted.

## Screenshots

*None.*

## Related Issue

https://issues.openmrs.org/browse/O3-602

## Other

*None.*

### CODE REVIEW

File: packages/esm-patient-registration-app/src/patient-registration/ui-components/overlay/index.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
```
Comment (manuelroemer) [inline packages/esm-patient-registration-app/src/patient-registration/ui-components/overlay/index.tsx:1]: One thing about the file: This overlay is only used inside the `IdField` right now. From my perspective, we could move this file out of the `ui-components` folder and into the folder of the `IdField` to improve discoverability (as long as the overlay isn't going to be used in other locations).

**Review by manuelroemer — COMMENTED** [review]
Thanks again for the whole work @vasharma05! There really is a large amount of work behind this! 🚀

To all readers: We have an interest in this PR being merged as soon as possible since it makes patient registration work again, i.e. it fixes the bug which caused registering a patient to fail and since the work in here is required for other tickets, i.e. this PR not being merged is a blocker for other issues.
This means that we'll try to resolve it quickly, especially since this state already _works_ very well (as seen in yesterday's squad call) - but nontheless, if there is important feedback, please feel free to comment here, even post-merge. Vineet will then look at it and create follow-up PRs as necessary.

<hr/>

@vasharma05 Regarding my own comments: I will leave it to you which ones you want to do in this PR and which ones you would like to do later in follow-up PRs. Some I already flagged as follow-up comments, but feel free to also move unflagged ones if they are too "big" for now.

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.component.tsx [hunk 2]
```diff
@@ -1,55 +1,82 @@
...
+            remove,
+            form: {
+              values: { identifiers },
+            },
+          }) => (
+            <>
+              {identifiers
+                .filter((identifier) => identifier.action !== 'DELETE')
+                .map((identifier: PatientIdentifierValue, index) => (
+                  <IdentifierInput key={index} index={index} patientIdentifier={identifier} remove={remove} />
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.component.tsx]: Yeah, but that was the case when the patient's identifier was fetched from the backend. If it's a new one, it won't have any uuid (undefined).

File: packages/esm-patient-registration-app/src/patient-registration/form-manager.ts [hunk 3]
```diff
@@ -138,41 +145,49 @@ export default class FormManager {
   }
 
   static getPatientIdentifiersToCreate(
-    values: FormValues,
-    patientUuidMap: object,
+    patientIdentifiers: PatientIdentifierValue[],
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/form-manager.ts]: Sure, I'll try to be consistent from the next time. Will do the changes in here too.
Thanks!

File: packages/esm-patient-registration-app/src/patient-registration/form-manager.ts [hunk 4]
```diff
@@ -79,15 +83,18 @@ export default class FormManager {
...
+      identifierTypes,
+      currentLocation,
+      abortController,
+    );
+
+    if (patientUuid) {
+      await Promise.all(
+        FormManager.savePatientIdentifiers(patientUuid, patientIdentifiers, values.identifiers, abortController),
+      );
+    }
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/form-manager.ts:97]: Yeah, I get you. I'll do the same in the next one.
Thanks!

**Review by FlorianRappl — APPROVED** [review]

File: packages/esm-patient-registration-app/src/patient-registration/input/custom-input/identifier/identifier-input.component.tsx [hunk 5]
```diff
@@ -1,106 +1,150 @@
...
+  index: number;
+  remove: <T>(index: number) => T;
 }
 
-export const IdentifierInput: React.FC<IdentifierInputProps> = ({ identifierType }) => {
-  const { validationSchema, setValidationSchema, setFieldValue } = React.useContext(PatientRegistrationContext);
-  const sources = identifierType.identifierSources;
-  const name = identifierType.fieldName;
+export const IdentifierInput: React.FC<IdentifierInputProps> = ({ patientIdentifier, index, remove }) => {
+  const { patientIdentifiers: identifierTypes } = useContext(ResourcesContext);
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/input/custom-input/identifier/identifier-input.component.tsx:19]: Well yes, I should've done that, would have spared me some time too.
DIdn't think of this, but will do.
Thanks!

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-types.tsx [hunk 6]
```diff
@@ -23,6 +23,19 @@ export interface FetchedPatientIdentifierType {
...
+  uuid?: string;
+  identifier: string;
+  identifierType?: string;
+  source: IdentifierSource;
+  autoGeneration?: boolean;
+  action: 'ADD' | 'UPDATE' | 'DELETE' | 'NONE';
+  // ADD -> add a new identifier
+  // UPDATE -> update an existing identifier
+  // DELETE -> delete an existing identifier
+  // NONE -> No action to be taken on the identifier
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-types.tsx:36]: I don't know about JSDoc but will learn and do the same in the next fixup commit.

[the remark this answers is not in the document]
Comment (manuelroemer) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-types.tsx:36]: That's basically just moving your comment in a block like this:

```ts
/**
 * your comment here
 */
```

Then editors like VS code will show you that comment when e.g. hovering over it with your mouse.

File: packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.component.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (manuelroemer) [reply packages/esm-patient-registration-app/src/patient-registration/field/id/id-field.component.tsx]: Ah, fair enough, then let's discard my comment!

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-types.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-types.tsx:36]: Thanks alot!

File: packages/esm-patient-registration-app/src/patient-registration/ui-components/identifier-selection-overlay/identifier-selection.scss [hunk 7]
```diff
@@ -0,0 +1,28 @@
+@import '../../patient-registration.scss';
+
+.overlayButtons {
+  display: grid;
+  grid-template-columns: 1fr 1fr;
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/ui-components/identifier-selection-overlay/identifier-selection.scss:5]: There's no variable for `fr`

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts [hunk 8]
```diff
@@ -168,3 +169,14 @@ export function getPhonePersonAttributeValueFromFhirPatient(patient: fhir.Patien
   }
   return result;
 }
+
+export async function getPatientIdentifiers(patientUuid: string): Promise<PatientIdentifierValue[]> {
+  const identifiers = await openmrsFetch<{ results: any[] }>(`/ws/rest/v1/patient/${patientUuid}/identifier?v=full`);
+  return identifiers.data.results.map((patientIdentifier) => ({
+    uuid: patientIdentifier.uuid,
+    identifier: patientIdentifier.identifier,
+    identifierType: patientIdentifier.identifierType.uuid,
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts:178]: I agree with changing the names of the function and the types, but regarding putting the patientIdentifier as a whole, It actually has a lot of values in it, which in turn will be saved with the Formik. I just kept the data, which was important with the form values, else, handling big data in the forms might decrease the performance of the Formik form.

What do you say?

File: packages/esm-patient-registration-app/src/patient-registration/ui-components/identifier-selection-overlay/identifier-selection.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (manuelroemer) [reply packages/esm-patient-registration-app/src/patient-registration/ui-components/identifier-selection-overlay/identifier-selection.scss:5]: You are absolutely right, my bad! 😖

**Pull request merged**